primitives/staking/src/offence.rs

```diff
@@ -209,3 +209,68 @@ pub struct OffenceDetails<Reporter, Offender> {
 	/// particular reporters.
 	pub reporters: Vec<Reporter>,
 }
+
+/// An abstract system to publish, check and process offence evidences.
+///
+/// Implementation details are left opaque and we don't assume any specific usage
+/// scenario for this trait at this level. The main goal is to group together some
+/// common actions required during a typical offence report flow.
+///
+/// Even though this trait doesn't assume too much, this is a general guideline
+/// for a typical usage scenario:
+///
+/// 1. An offence is detected and an evidence is submitted on-chain via the
+///    [`OffenceReportSystem::publish_evidence`] method. This will construct
+///    and submit an extrinsic transaction containing the offence evidence.
+///
+/// 2. If the extrinsic is unsigned then the transaction receiver may want to
+///    perform some preliminary checks before further processing. This is a good
+///    place to call the [`OffenceReportSystem::check_evidence`] method.
+///
+/// 3. Finally the report extrinsic is executed on-chain. This is where the user
+///    calls the [`OffenceReportSystem::process_evidence`] to consume the offence
+///    report and enact any required action.
+pub trait OffenceReportSystem<Reporter, Evidence> {
+	/// Longevity, in blocks, for the evidence report validity.
+	///
+	/// For example, when using the staking pallet this should be set equal
+	/// to the bonding durationin blocks, not eras).
+	type Longevity: Get<u64>;
+
+	/// Publish an offence evidence.
+	///
+	/// Common usage: submit the evidence on-chain via some kind of extrinsic.
+	fn publish_evidence(evidence: Evidence) -> Result<(), ()>;
```

Comment:
nit: How about renaming it to `publish_evidence_extrinsic/transaction`? Or maybe even `sumbit`?

Reply:
You are right that can be more clear in the current usage context (and maybe the only one context for this trait 😃) .
But after thinking about it I realized (indeed since our last call I've updated the docs trying to be more clear as possible) that the trait wants to be something used for any generic system that is able to publish/check/digest and offence (usually in this sequence).

I was hesitant about the "submit" word because it makes me think about submitting an extrinsic/transaction. and that is not the only way the subsystem can be used for (at least at a design level 😅).

But maybe yeah. I can rename it to `submit` instead of publish...

Reply:
just to elaborate a bit more. I was thinking about something like one implementor of this trait may (for example) create its report system that is not on-chain and just publish the offences somewhere (e.g. on a website or a centralized db). The trait don't want to enter in such details

Reply:
I don't have a too strong opinion on this, but notice that this method still needs to be called from some pallet-exposed method. In our case, for all implementors of this trait, that method is named `submit_unsigned_equivocation_report`, so I think in the cases where it is used to actually submit a transaction it is already explicit.